[Solidity] code refactoring, bug fixed for inheritance and more

ChenfengWei0 opened the pull request.

ChenfengWei1 committed:
[siolidity] partially fix the ctor

ChenfengWei1 committed:
[siolidity] partially fix the ctor-2

ChenfengWei1 committed:
[siolidity] fix new expression

ChenfengWei1 committed:
[siolidity] add regression tests

ChenfengWei1 committed:
[siolidity] fix contractmembercall

ChenfengWei1 committed:
[siolidity] add regression tests

ChenfengWei1 committed:
[siolidity] add regression tests

ChenfengWei1 committed:
[siolidity] major code refactoring: to cpp format

ChenfengWei1 committed:
[siolidity] fix static lifetime

ChenfengWei1 committed:
[siolidity] fix global variable

ChenfengWei1 committed:
[siolidity] add this ref for state variable

ChenfengWei1 committed:
[siolidity] fix constructor initialization

ChenfengWei1 committed:
[siolidity] fix func_call this ptr

ChenfengWei1 committed:
[siolidity] bug fix

ChenfengWei1 committed:
[siolidity] refactoring

ChenfengWei1 committed:
[siolidity] refactoring

ChenfengWei1 committed:
[siolidity] refactoring

ChenfengWei1 committed:
[siolidity] fix inheritance

ChenfengWei1 committed:
[siolidity] fix string & array

ChenfengWei1 committed:
[siolidity] fix tuple

ChenfengWei1 committed:
[siolidity] fix func_call && regression tests

ChenfengWei1 committed:
[siolidity] fix tuple

ChenfengWei1 committed:
[siolidity] fix regression tests

ChenfengWei1 committed:
[siolidity] rm testing

intrigus-lgtm commented on src/solidity-frontend/solidity_convert.cpp:
```diff
@@ -531,7 +621,30 @@ bool solidity_convertert::get_var_decl(
+    t.identifier().as_string() == prefix + current_contractName)
+  {
+    // contract Base{
+    //   Base x;
```
How is this implemented/why does it work in Solidity?

ChenfengWei0 replied:
```solidity
contract Base
{
    Base x;
    uint public y;
    constructor()
    {
        y = 1;
    }

    function test() public   {
        x = Base(0x9cF86D8D08bC34248210474C4B019befb0fE70fA);
        assert(x.y() == 1); 
    }
}
```
- We first deploy this contract, gather the address 
- We initialize x with the contract address and deploy it again.
- Run function `test` and the assertion holds.

fbrausse commented on src/pointer-analysis/dereference.cpp:
```diff
@@ -908,10 +908,27 @@ void dereferencet::build_reference_rec(
+        "Can't construct rvalue reference to array type during dereference\n"
+        "(It isn't allowed by C anyway)\n");
+      abort();
+    }
```
This fixes nothing.

ChenfengWei0 replied:
It just avoid the abort. The result is correct.

fbrausse replied:
What is the result?

ChenfengWei0 replied:
E.g.

```solidity
contract test {
    uint[3] public t = [0, 0]; // t = [0,0,0]

    constructor() {
        (t, t) = ([1], [2]);  // t = [2,0,0]; t=[1,0,0]
    }
}

contract test1 is test {
    function test2() public view {
        assert(t[0] == 1);
        assert(t[1] == 0);
        assert(t[2] == 0);
    }
}
```
- Command: `--unwind 10 --no-unwinding-assertions`
- Result: `VERIFICATION SUCCESSFUL`

I am not sure what's going on in the middleware & backend. 
E.g.
```solidity
 int x[3] = [1, 2, 3];
 constructor(){
     x = [4, 5, 6];
 }
```
I guess the assignment will just overwrite the x's value. The corresponding symbol table is:
```
Symbol......: sol:@C@test@F@test#19
Module......: example
Base name...: test
Mode........: C++
Type........: constructor  (struct test *)
Value.......:
{
&(this->t = { 1, 2, 3 })[0];

&(this->t = { 4, 5, 6 })[0];

}
Flags.......: lvalue
Location....: file example.sol line 4
```

fbrausse replied:
Thanks for the examples, they're helpful. The reason the dereference module doesn't handle array dereferences is because there are no C semantics for it - since C doesn't have that construct: A dereference in C always refers to either a built-in type like `int` or to a non-array struct or union (member). For the C++ frontend we got around that problem in two ways:
1. array-typed statements: https://github.com/esbmc/esbmc/blob/7b1e7f2d750f8d579db145a118338a9c52b81347/src/clang-c-frontend/clang_c_adjust_code.cpp#L45-L60
2. during a constructor call: https://github.com/esbmc/esbmc/blob/7b1e7f2d750f8d579db145a118338a9c52b81347/src/clang-cpp-frontend/clang_cpp_convert.cpp#L1184-L1190

Would that be an option for you with Solidity?

ChenfengWei0 replied:
Could they support the dynamic array? Like:
```solidity
unit[] x;

constructor()
{
  x = [1];
}

function test() public
{
  x = [1, 2, 3];
  x = [4, 5];
}
```

---

I can make the fixed array to pointer type in the front end, like the dynamic array. E.g. 
```
Symbol......: sol:@C@Base@x#3
Module......: example
Base name...: x
Mode........: C++
Type........: unsigned _ExtInt(256) *
Value.......: 0
Flags.......: lvalue file_local
Location....: file example.sol line 2

Symbol......: sol:@C@Base@F@test#20
Module......: example
Base name...: test
Mode........: C++
Type........: void (struct Base *)
Value.......:
{
this->x = &{ 1, 2, 3 }[0];

this->x = &{ 4, 5 }[0];

}
Flags.......: lvalue
Location....: file example.sol line 4
```
But I don't feel like changing the code for the fixed array as they already 'work' somehow.

ChenfengWei1 committed:
[siolidity] rm testing

ChenfengWei1 committed:
[solidity] fix mapping

ChenfengWei1 committed:
[solidity] fix mapping index

ChenfengWei1 committed:
[solidity] fix algo

ChenfengWei1 committed:
[solidity] fix contract fields&methods

ChenfengWei1 committed:
[solidity] partially fix alignment

ChenfengWei1 committed:
[solidity] partially fix array

ChenfengWei1 committed:
[solidity] fix array

ChenfengWei1 committed:
[solidity] fix string type

ChenfengWei1 committed:
[solidity] bug fix

ChenfengWei1 committed:
[solidity] temporary fix

ChenfengWei1 committed:
[solidity] fix dyn array && string

ChenfengWei1 committed:
[solidity] bug fix

ChenfengWei1 committed:
[solidity] refactoring

ChenfengWei0 committed:
Apply automatic changes

ChenfengWei1 committed:
[solidity] bug fix

ChenfengWei1 committed:
[solidity] fix string type

ChenfengWei1 committed:
[solidity] fix string type

ChenfengWei1 committed:
[solidity] fix string parameters

ChenfengWei1 committed:
[solidity] fix inheritance

ChenfengWei1 committed:
[solidity] fix mapping

ChenfengWei1 committed:
[solidity] fix ctor

ChenfengWei1 committed:
[solidity] fix inheritance

ChenfengWei1 committed:
[solidity] fix regression

ChenfengWei1 committed:
[solidity] implement library function

ChenfengWei1 committed:
[solidity] add TODO

ChenfengWei1 committed:
[solidity] fix library function call

ChenfengWei1 committed:
[solidity] rebase

ChenfengWei1 committed:
[solidity] fix rev

ChenfengWei1 committed:
[solidity] fix dyn array && update regression test

ChenfengWei1 committed:
[solidity] fix mapping

ChenfengWei1 committed:
[solidity] fix initializer

ChenfengWei1 committed:
[solidity] fix fixed size array

ChenfengWei1 committed:
[solidity] fix regression

ChenfengWei1 committed:
[solidity] fix tuple

ChenfengWei1 committed:
[solidity] add regression

ChenfengWei0 committed:
Apply automatic changes

ChenfengWei1 committed:
[solidity] rollback changes

ChenfengWei0 committed:
Apply automatic changes

ChenfengWei1 committed:
[solidity] switch between _extint and _bigint

ChenfengWei1 committed:
[solidity] rm assert

ChenfengWei0 commented:
This PR introduces a major refactoring, shifting from a C-based implementation to a C++-based one. This transition enables better support for complex structures such as inheritance (the previous version was somewhat broken), contract instances (which is virtually identical to class instances), and more. Additionally, it includes several bug fixes and test improvements.

- Major refactoring: Migrated from C-based to C++-based architecture to support advanced features like inheritance and contract instances.
- Bug fixes: Addressed issues in features including constructors, dynamic arrays, tuples, and mappings.
- Test improvements: Fixed or removed some regression tests to align with the new implementation.

lucasccordeiro committed:
Merge branch 'master' into solidity_2

lucasccordeiro commented:
Thanks for submitting this PR, @ChenfengWei0.
